### PR TITLE
EnTp, 1 nonmatching, and object_tp

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Had a sting in the tail, tail of two functions, etc...

All functions matched apart from Init, which cannot be persuaded to put zero in a temp, the main reason it took so long to decomp and document. I've documented what I had a reasonable understanding of, and left a few comments where I couldn't work it out; this actor is surprisingly complicated for such a simple-seeming enemy. As ever, naming suggestions welcome if names are inappropriate or unclear. Also did the object.

Weird actor, lots of loops to deal with the tail, and some very confusing variables in the struct. Shows clear signs of being an early actor based on the functions it uses. 

Also corrected a couple of spellings: "skulltulla", and "DrawBridge" (partly because it's one word, partly because it's very confusing to have what appears to be the word Draw used for naming functions that aren't Draws lol).

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c [hunk 1]
```diff
@@ -141,15 +141,15 @@ void BgSpot00Hanebasi_Destroy(Actor* thisx, GlobalContext* globalCtx) {
...
 
-void BgSpot00Hanebasi_DrawBridgeWait(BgSpot00Hanebasi* this, GlobalContext* globalCtx) {
+void BgSpot00Hanebasi_DrawbridgeWait(BgSpot00Hanebasi* this, GlobalContext* globalCtx) {
     BgSpot00Hanebasi* child = (BgSpot00Hanebasi*)this->dyna.actor.child;
 
     if ((gSaveContext.sceneSetupIndex >= 4) || (!(CHECK_QUEST_ITEM(QUEST_KOKIRI_EMERALD))) ||
         (!(CHECK_QUEST_ITEM(QUEST_GORON_RUBY))) || (!(CHECK_QUEST_ITEM(QUEST_ZORA_SAPPHIRE))) ||
         (gSaveContext.eventChkInf[8] & 1)) {
         if (this->dyna.actor.shape.rot.x != 0) {
             if (Flags_GetEnv(globalCtx, 0) || ((gSaveContext.sceneSetupIndex < 4) && (gSaveContext.nightFlag == 0))) {
```
Comment (louist103) [inline src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c:151]:
```suggestion
        if (this->dyna.actor.shape.rot.x != 0) && ((Flags_GetEnv(globalCtx, 0) || ((gSaveContext.sceneSetupIndex < 4) && (gSaveContext.nightFlag == 0))) {
```
While you are in here can this work? I probably got the if wrong but you get the idea.

File: src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c [hunk 2]
```diff
@@ -141,15 +141,15 @@ void BgSpot00Hanebasi_Destroy(Actor* thisx, GlobalContext* globalCtx) {
...
-                this->actionFunc = BgSpot00Hanebasi_DrawBridgeRiseAndFall;
+                this->actionFunc = BgSpot00Hanebasi_DrawbridgeRiseAndFall;
                 this->destAngle = 0;
                 child->destAngle = 0;
                 return;
```
Comment (louist103) [inline src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c:155]:
```suggestion
```
Also this can probably go if you adjust the `if`s.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 3]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
     1,
     D_80B22A90,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/func_80B20DE0.s")
 
+typedef enum {
+    /* 00 */ TAILPASARAN_DMGEFF_NONE, // No effect
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:84]:
```suggestion
    /* 00 */ TAILPASARAN_DMGEFF_NONE,
```
The comment probably isn't necessary.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 4]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    /* Unknown 1     */ DMG_ENTRY(0, TAILPASARAN_DMGEFF_NONE),
+    /* Unblockable   */ DMG_ENTRY(0, TAILPASARAN_DMGEFF_NONE),
+    /* Hammer jump   */ DMG_ENTRY(4, TAILPASARAN_DMGEFF_SHOCKING),
+    /* Unknown 2     */ DMG_ENTRY(0, TAILPASARAN_DMGEFF_NONE),
+};
+
+// sInitChain
+static InitChainEntry D_80B22AE4[] = {
+    ICHAIN_F32(targetArrowOffset, 10, ICHAIN_STOP),
+};
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:128]:
```suggestion
static InitChainEntry sInitChain[] = {
    ICHAIN_F32(targetArrowOffset, 10, ICHAIN_STOP),
};
```
Also fix in the assembly file.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 5]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/func_80B20DE0.s")
 
+typedef enum {
+    /* 00 */ TAILPASARAN_DMGEFF_NONE, // No effect
+    /* 01 */ TAILPASARAN_DMGEFF_DEKUNUT,
+    /* 14 */ TAILPASARAN_DMGEFF_SHOCKING = 14, // Kills the Tailpasaran but shocks Player
+    /* 15 */ TAILPASARAN_DMGEFF_INSULATING     // Kills the Tailpasaran and does not shock Player
+} TailpasaranDamageEffect;
+
+static DamageTable D_80B22AC4 = {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:90]:
```suggestion
static DamageTable sDamageTable= {
```
Also fix in the assembly file.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 6]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+#ifdef NON_MATCHING
+// Regalloc, some stack, and compiler refuses to put zero in $s4 with temp_s4; all seems to be related
+void EnTp_Init(Actor* thisx, GlobalContext* globalCtx2) {
+    GlobalContext* globalCtx = globalCtx2;
+    EnTp* this = THIS;
+
+    EnTp* now;
+    EnTp* next;
+    s32 i;
+    // s64 new_var;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:143]:
```suggestion
```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 7]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    s64 temp_s4;
+
+    Actor_ProcessInitChain(&this->actor, D_80B22AE4);
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:146]:
```suggestion
    Actor_ProcessInitChain(&this->actor, sInitChain);
```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 8]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    this->actor.targetMode = 3;
+    this->actor.colChkInfo.damageTable = &D_80B22AC4;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:148]:
```suggestion
    this->actor.colChkInfo.damageTable = &sDamageTable;
```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 9]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+        this->collider.base.acFlags |= AC_HARD;
+        this->collider.elements->dim.modelSphere.radius = this->collider.elements->dim.worldSphere.radius = 8;
+        EnTp_Head_SetupWait(this);
+        this->actor.focus.pos = this->actor.world.pos;
+        this->actor.flags |= 0x15;
+        Actor_SetScale(&this->actor, 1.5f);
+
+        for (i = 0; i <= 6; i++) {
+            temp_s4 = 0;
+            // new_var = 0;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:169]:
```suggestion
```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 10]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+}
+
+void EnTp_Head_BurrowReturnHome(EnTp* this, GlobalContext* globalCtx) {
+    static Vec3f bubbleAccel = { 0.0f, -0.5f, 0.0f };
+    static Color_RGBA8 bubblePrimColor = { 255, 255, 255, 255 };
+    static Color_RGBA8 bubbleEnvColor = { 150, 150, 150, 0 };
+    Vec3f bubbleVelocity;
+    Vec3f bubblePos;
+    s32 closeToFloor;
+    EnTp* now;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:508]:
```suggestion
    EnTp* child;
```
Or something similar.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 11]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+void EnTp_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    EnTp* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/func_80B21B90.s")
+    OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_tp.c", 1451);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/func_80B21BDC.s")
+    if (this->unk_150 != 2) {
+        if ((thisx->params <= TAILPASARAN_HEAD) || (thisx->params == TAILPASARAN_HEAD_DYING)) {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:746]:
```suggestion
        if ((this->actor.params <= TAILPASARAN_HEAD) || (this->actor.params == TAILPASARAN_HEAD_DYING)) {
```
If possible.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 12]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+                      G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+            gSPDisplayList(POLY_XLU_DISP++, gTailpasaranTailSegmentDL);
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/EnTp_Update.s")
+    CLOSE_DISPS(globalCtx->state.gfxCtx, "../z_en_tp.c", 1495);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Tp/EnTp_Draw.s")
+    if ((thisx->params <= TAILPASARAN_TAIL) || (thisx->params == TAILPASARAN_TAIL_DYING)) {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:774]:
```suggestion
    if ((this->actor.params <= TAILPASARAN_TAIL) || (this->actor.params == TAILPASARAN_TAIL_DYING)) {
```
Again if possible.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 13]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    Actor_ProcessInitChain(&this->actor, D_80B22AE4);
+    this->actor.targetMode = 3;
+    this->actor.colChkInfo.damageTable = &D_80B22AC4;
+    ActorShape_Init(&this->actor.shape, 0.0f, ActorShadow_DrawCircle, 0.14f);
+    this->unk_150 = 0;
+    this->actor.colChkInfo.health = 1;
+    now = this;
+    this->alpha = 255;
+    Collider_InitJntSph(globalCtx, &this->collider);
+    Collider_SetJntSph(globalCtx, &this->collider, &this->actor, &D_80B22AB4, this->colliderItems);
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:155]:
```suggestion
    Collider_SetJntSph(globalCtx, &this->collider, &this->actor, &sJntSphInit, this->colliderItems);
```

**Review by louist103 — CHANGES_REQUESTED** [review]
Nice job on this.  Just some styling things.

File: src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c (hunk 1 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c:151]: This does weird things with the regalloc: notice the `if (this) {}` is after the return

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 10 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:508]: I'm not sure about this, since it iterates over the generations of children

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 11 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:746]: It is required.

File: src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c [hunk 14]
```diff
@@ -22,8 +22,8 @@ void BgSpot00Hanebasi_Destroy(Actor* thisx, GlobalContext* globalCtx);
 void BgSpot00Hanebasi_Update(Actor* thisx, GlobalContext* globalCtx);
 void BgSpot00Hanebasi_Draw(Actor* thisx, GlobalContext* globalCtx);
 
-void BgSpot00Hanebasi_DrawBridgeWait(BgSpot00Hanebasi* this, GlobalContext* globalCtx);
-void BgSpot00Hanebasi_DrawBridgeRiseAndFall(BgSpot00Hanebasi* this, GlobalContext* globalCtx);
+void BgSpot00Hanebasi_DrawbridgeWait(BgSpot00Hanebasi* this, GlobalContext* globalCtx);
+void BgSpot00Hanebasi_DrawbridgeRiseAndFall(BgSpot00Hanebasi* this, GlobalContext* globalCtx);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c:26]: Why did you change the capitalization here?

File: assets/xml/objects/object_tp.xml [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+        <DList Name="gTailpasaranHeadDL" Offset="0x08D0"/>
+
+        <!-- Tailpasaran textures -->
+        <Texture Name="gTailpasaranHeadRearTex" OutName="tailpasaran_head_rear" Format="rgb5a1" Width="8" Height="16" Offset="0x08E8"/>
+        <Texture Name="gTailpasaranHeadClawBaseTex" OutName="tailpasaran_head_claw_base" Format="rgb5a1" Width="8" Height="8" Offset="0x09E8"/>
+        <Texture Name="gTailpasaranHeadClawTex" OutName="tailpasaran_head_claw" Format="rgb5a1" Width="16" Height="16" Offset="0x0A68"/>
+        <Texture Name="gTailpasaranTailSegmentTex" OutName="tailpasaran_tail_segment" Format="ia8" Width="64" Height="64" Offset="0x0C68"/>
+        <Texture Name="gTailpasaranHeadMouthpartsTex" OutName="tailpasaran_head_mouthparts" Format="rgb5a1" Width="8" Height="8" Offset="0x1C68"/>
+
+
```
Comment (AngheloAlf) [inline assets/xml/objects/object_tp.xml:17]:
```suggestion

```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 16]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+                                 0, 255, 1, 9, 1);
+
+            effectPos.x = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.x;
+            effectPos.z = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.z;
+            effectPos.y = ((Rand_ZeroOne() - 0.5f) * 5.0f) + this->actor.world.pos.y;
+            EffectSsDeadDb_Spawn(globalCtx, &effectPos, &effectVelAccel, &effectVelAccel, 100, 0, 255, 255, 255, 255, 0,
+                                 0, 255, 1, 9, 1);
+            Item_DropCollectibleRandom(globalCtx, &this->actor, &this->actor.world.pos, 0x50);
+        } else {
+            for (i = 0; i < 1; i++) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:331]: Really?

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 17]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    if (this->red != 0) {
+        this->red -= 15;
+    }
+
+    if (Math_CosF(this->heightPhase) == 0.0f) {
+        this->extraHeightVariation = Rand_ZeroOne() * 4.0f;
+    }
+
+    this->actor.world.pos.y +=
+        (Math_CosF(this->heightPhase) * ((this->actor.speedXZ * 0.25f) + this->extraHeightVariation));
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:411]:
```suggestion
        Math_CosF(this->heightPhase) * (this->actor.speedXZ * 0.25f + this->extraHeightVariation);
```

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 10 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:508]: I don't think `now` is a good name for it. `child` isn't a good alternative either imo.

**Review by AngheloAlf — COMMENTED** [review]
Looks nice. Just a few minor comments.

File: src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c (hunk 14 above)
Comment (fig02) [reply src/overlays/actors/ovl_Bg_Spot00_Hanebasi/z_bg_spot00_hanebasi.c:26]: drawbridge is one word

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c [hunk 18]
```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
...
+    this->alpha -= 20;
+
+    if (this->alpha < 20) {
+        this->alpha = 0;
+        Actor_Kill(&this->actor);
+    }
+}
+
+void EnTp_Head_SetupTakeOff(EnTp* this) {
+    this->timer = (s32)((Rand_ZeroOne() * 15.0f) + 40.0f);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Tp/z_en_tp.c:377]: Is this cast necessary?

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 4 above)
Comment (Roman971) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:128]: Yeah I agree about updating the names for these static variables, since it should work fine now

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 18 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:377]: I thought it might be, since timer is an s16, but apparently not.

File: src/overlays/actors/ovl_En_Tp/z_en_tp.c (hunk 10 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Tp/z_en_tp.c:506]: Yeah, happy for suggestion of alternatives. It's used to iterate over the tail segments, which are linked using parent-child, ending at the head.

**Review by fig02 — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]
